design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,398 @@
+
+The *prefix* and *gateway* parameters can be given for each pool of the list,
+or globally. If they are given for a pool they will override the global
+settings, that are default values. The *prefix* and *gateway* will be set on
+the *IPAddress* and can be fetched from a *Template*.
+
+The *allocations* fields is a map of object name and ip address that allow a
+user to specify a set of static allocations for some objects.
+
+The *namePrefix* contains the prefix used to name the IPAddress objects
+created. It must remain the same for a subnet, across updates or changes in the
+IPPool object to keep the existing leases.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update. In case of an error during the allocation (pool exhaustion for example),
+the error would be reported on the Consumer object, the *error* boolean and
+*errorMessage* field on the *RenderedData* object.
+The *allocations* map will map the IP address to the *RenderedData* object it was
+allocated for and the *addresses* will map the *RenderedData* objects with the
+*IPAddress* objects.
+
+The *IPAddress* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPPool
+    name: pool-1
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPClaim
+    name: RenderedData-1
+spec:
+  ipClaim:
+    Name: RenderedData-1
+  Address: 192.168.0.11
+  prefix: 24
+  gateway: 192.168.0.1
+  IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+The *IPAddress* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPClaim
+metadata:
+  name: RenderedData-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: metadata.metal3.io/v1alpha1
+    kind: Data
+    name: data-1
+spec:
+  owner:
+    name: RenderedData-1
+  ipPool:
+    name: pool-1
+status:
+  ready: true
+  ipAddress:
+    name: pool-1-192-168-0-11
+  error: false
+  errorMessage: ""
+```
+
+For each *IPClaim*, the controller reconciling the *IPCalim* will select an
+available IP address randomly from the *IPPool*, if the object name is not in
+the *allocations* map in the object *spec* or in the *addresses* map in the
+*status*, in that case it would select that IP address. An error would be
+reflected on the claim.
+
+Once the IP address is selected, the controller will create an *IPAddress*
+for that address. In case of conflict, it will list all the *IPAddress*
+objects that have an owner reference to this *IPPool* and update the
+status with the mapping of IP addresses and *IPPool* object names. It will
+then randomly select an available IP address. Once the *IPAddress* object
+is created, the *IPPool* object status will be updated with the new map. Then
+The *IPClaim* status will be updated with the *IPAddress* reference and the
+*ready* field set to true.
```

Comment:
On second thought, maybe the pool's status field isn't any more likely to produce collisions than looking at the IPAddress resources.

Reply:
I think using the CRs directly as source of truth instead of a cached version from the status would be better, with a simpler logic wrt. this status update. However, I'd keep the status in place, as it is a single location where the user can get the overall view on the allocations. Do you mind if I directly put your explanation of the algorithm in the PR ?

Reply:
If we only use the status for the user's sake, then we can remove some synchronization and update logic, making the allocation process more uniform and stream-lined. I'll give it a try.

Reply:
It's fine to use my wording if that helps.